### PR TITLE
fixes #5178 - unify API parameters and return values. User creation should not require payload wrapped with 'user' root

### PR DESCRIPTION
**Pull request by isratrade**

### CODE REVIEW

File: app/controllers/api/v2/users_controller.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 module Api
   module V2
     class UsersController < V2::BaseController
+
+      wrap_parameters User, :include => (User.attribute_names + ['password'])
```
Comment (isratrade) [inline app/controllers/api/v2/users_controller.rb:5]: it seems that filtered attributes are not wrapped automatically.  This was forcing the user to wrap the payload with root load 'user' to create the object.
https://www.bountysource.com/issues/336019-password-attribute-not-wrapped-by-default-when-using-has_secure_password

File: app/controllers/api/v2/users_controller.rb [hunk 2]
```diff
@@ -46,11 +49,8 @@ def show
       param_group :user, :as => :create
 
       def create
-        if @user.save
-          process_success
-        else
-          process_resource_error
-        end
+        @user = User.new(params[:user])
+        process_response @user.save
```
Comment (isratrade) [inline app/controllers/api/v2/users_controller.rb]: this is just a clean up as all the other v2 controllers #create use this notation.

File: app/controllers/api/v2/users_controller.rb (hunk 1 above)
Comment (thomasmckay) [inline app/controllers/api/v2/users_controller.rb:5]: Good find!

**Comment by domcleal** [comment]
I think we could have some more instances of this throughout the codebase, though it's going to be hard to track down.  As an example, I tried creating an OpenStack CR and found that both the provider and tenant attributes don't get wrapped, so POSTing an unwrapped set of JSON, you're told that the provider is missing.

A quick search for "def \S+=" and attr_accessor/writer shows quite a few possibilities for attributes we'd need to add, with most of them in the CR models.  I think we might want to define something in the model to return all of the attributes that should be wrapped, for easy reference from the controllers.

File: app/controllers/api/v2/users_controller.rb [hunk 3]
```diff
@@ -1,6 +1,9 @@
 module Api
   module V2
     class UsersController < V2::BaseController
```
Comment (domcleal) [inline app/controllers/api/v2/users_controller.rb:3]: v1 as well?

**Comment by thomasmckay** [comment]
confirmed fixes users v2 api create
does not fix users v2 api update

**Comment by isratrade** [comment]
@thomasmckay, what is the error that it doesn't work with users v2 api update?

**Comment by isratrade** [comment]
@domcleal, I added the `attr_accessor` attributes as well as other setter methods "def \S+=" 
I thought to pro-grammatically add add setter methods such as 

``` ruby
klass_instance.methods.select {|m| m.to_s.end_with?('=') }
```

but @ohadlevy was not in favor of adding no many rails public methods to the API. He suggests (and I agree) to go with the strong_parameters gem, already entered as http://projects.theforeman.org/issues/3917

**Comment by komidore64** [comment]
can we get some priority on this?

**Comment by isratrade** [comment]
@ohadlevy, do you feel comfortable to merge this as is and deal with strong parameters in a separate issue http://projects.theforeman.org/issues/3917

File: app/controllers/api/v2/hosts_controller.rb [hunk 4]
```diff
@@ -2,6 +2,8 @@ module Api
   module V2
     class HostsController < V2::BaseController
 
+      wrap_parameters Host::Base, :include => (Host::Base.attribute_names + ['image_file', 'lookup_values_attributes', 'is_owned_by', 'overwrite', 'progress_report_id'])
```
Comment (ohadlevy) [inline app/controllers/api/v2/hosts_controller.rb]: imho lookup_value_attributes should not be used at all over API, they are for nested attributes.

**Comment by ohadlevy** [comment]
@isratrade agreed.

File: app/controllers/api/v2/hosts_controller.rb (hunk 4 above)
Comment (isratrade) [inline app/controllers/api/v2/hosts_controller.rb]: right, in any case, the API renames them to `override_values`

**Comment by isratrade** [comment]
rebased after removing `lookup_value_attributes`

**Comment by lzap** [comment]
Merged as @ohadlevy agreed to split these into two PRs.

Commit is c980e9a, thanks.

**Pull request closed without merging**